src/game/client/shadow/w3dvolumetricshadow.cpp

```diff
@@ -0,0 +1,512 @@
+
+        RenderObjClass *lod_model = hlod->Peek_Lod_Model(lod_index, model_index);
+
+        if (lod_model != nullptr) {
+
+            if (lod_model->Class_ID() == RenderObjClass::CLASSID_MESH) {
+
+                captainslog_dbgassert(m_meshCount < MAX_SHADOW_CASTER_MESHES, "Too many shadow sub-meshes");
+
+                MeshClass *mesh = (MeshClass *)lod_model;
+
+                geo_mesh->m_mesh = mesh;
+                geo_mesh->m_modelIndex = model_index;
+
+                MeshModelClass *model = geo_mesh->m_mesh->Peek_Model();
+
+                if (geo_mesh->m_mesh->Is_Alpha() || geo_mesh->m_mesh->Is_Translucent()) {
+
+                    if (!model->Get_Flag(MeshGeometryClass::CAST_SHADOW)) {
+                        continue;
+                    }
+                }
+
+                // TODO Look into. Is this why infantry can't have shadows?
+                if (model->Get_Flag(MeshGeometryClass::SKIN)) {
+                    continue;
+                }
+
+                geo_mesh->m_numVerts = model->Get_Vertex_Count();
+                geo_mesh->m_verts = model->Get_Vertex_Array();
+                geo_mesh->m_numPolygons = model->Get_Polygon_Count();
+                geo_mesh->m_polygons = model->Get_Polygon_Array();
+
+                if (geo_mesh->m_numVerts > 16384) {
+                    return 0;
+                }
+
+                unsigned short verts[16384];
+                memset(verts, -1, sizeof(verts));
+
+                int num = geo_mesh->m_numVerts;
+
+                for (int i = 0; i < geo_mesh->m_numVerts; ++i) {
+                    if (verts[i] == 0xFFFF) {
+
+                        for (int j = i + 1; j < geo_mesh->m_numVerts; ++j) {
+
+                            Vector3 d = geo_mesh->m_verts[i] - geo_mesh->m_verts[j];
+
+                            if (d.Length2() == 0.0f) {
+                                verts[j] = i;
+                                --num;
+                            }
+                        }
+                        verts[i] = i;
+                    }
+                }
+
+                geo_mesh->m_parentVerts = new unsigned short[geo_mesh->m_numVerts];
+                memcpy(geo_mesh->m_parentVerts, verts, 2 * geo_mesh->m_numVerts);
+                geo_mesh->m_numVerts = num;
+                m_numTotalsVerts += num;
+                geo_mesh->m_parentGeometry = this;
+
+                ++geo_mesh;
+
+                ++m_meshCount;
+            }
+        }
+
+        if (lod_model != nullptr) {
+            if (lod_model->Class_ID() == RenderObjClass::CLASSID_SHDMESH) {
+                captainslog_dbgassert(0, "Meshes using shaders are not supported! How did you end up here!?");
+            }
+        }
+    }
+    return m_meshCount != 0;
+}
+
+int W3DShadowGeometry::Init_From_Mesh(RenderObjClass *robj)
+{
+
+    W3DShadowGeometryMesh *geo_mesh = &m_meshList[m_meshCount];
```

Comment:
This is duplicating code almost identical to the line above suggest we deduplicate.

Reply:
marking that as a todo, bfme2 does provide the how but not sure if its the same still